### PR TITLE
implement `dpnp.einsum`

### PR DESCRIPTION
**Pull request by vtavana**
In this PR, `dpnp.einsum` is implemented and `dpnp.rinsum_path` falls back on `NumPy` since `dpnp_ndarray` are not used in it.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 1]
```diff
@@ -153,6 +545,199 @@ def _gemm_batch_matmul(exec_q, x1, x2, res, x1_is_2D, x2_is_2D, dev_tasks_list):
...
+    >>> _greedy_path(isets, oset, idx_sizes, 5000)
+    [(0, 2), (0, 1)]
+
+    """
+
+    # Handle trivial cases that leaked through
+    if len(input_sets) == 1:
+        return [(0,)]
+    elif len(input_sets) == 2:
+        return [(0, 1)]
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:591]: not covered by any test

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 2]
```diff
@@ -210,6 +795,340 @@ def _op_res_dtype(*arrays, dtype, casting, sycl_queue):
...
+                )
+                new_pos = positions + [con]
+                iter_results.append((total_cost, new_pos, new_input_sets))
+
+        # Update combinatorial list, if we did not find anything return best
+        # path + remaining contractions
+        if iter_results:
+            full_results = iter_results
+        else:
+            path = min(full_results, key=lambda x: x[0])[1]
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:864]: missed to coverage

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 3]
```diff
@@ -210,6 +795,340 @@ def _op_res_dtype(*arrays, dtype, casting, sycl_queue):
...
+            path += [tuple(range(len(input_sets) - iteration))]
+            return path
+
+    # If we have not found anything return single einsum contraction
+    if len(full_results) == 0:
+        return [tuple(range(len(input_sets)))]
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:870]: missed to coverage

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 4]
```diff
@@ -210,6 +795,340 @@ def _op_res_dtype(*arrays, dtype, casting, sycl_queue):
...
+    removed_size = sum(old_sizes) - new_size
+
+    # NB: removed_size used to be just the size of any removed indices i.e.:
+    #     helpers.compute_size_by_dict(idx_removed, idx_dict)
+    cost = _flop_count(idx_contract, idx_removed, len(positions), idx_dict)
+    sort = (-removed_size, cost)
+
+    # Sieve based on total cost as well
+    if (path_cost + cost) > naive_cost:
+        return None
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:1126]: missed to coverage

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 5]
```diff
@@ -476,6 +1554,215 @@ def dpnp_cross(a, b, cp, exec_q):
...
+        tmp_subscripts = set(itertools.chain.from_iterable(input_subscripts))
+        for label in output_subscript:
+            if label not in tmp_subscripts:
+                raise ValueError(
+                    "einstein sum subscripts string included output subscript "
+                    f"'{_chr(label)}' which never appeared in an input."
+                )
+        if len(output_subscript) != len(set(output_subscript)):
+            for label in output_subscript:
+                if output_subscript.count(label) >= 2:
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:1635]: missed to coverage

File: dpnp/dpnp_iface_linearalgebra.py [hunk 6]
```diff
@@ -166,15 +167,59 @@ def dot(a, b, out=None):
...
+    Parameters
+    ----------
+    subscripts : str
+        Specifies the subscripts for summation as comma separated list of
+        subscript labels. An implicit (classical Einstein summation)
+        calculation is performed unless the explicit indicator '->' is
+        included as well as subscript labels of the precise output form.
+    *operands : sequence of {dpnp.ndarrays, usm_ndarray}
+        These are the arrays for the operation.
+    out : {dpnp.ndarrays, usm_ndarray}, optional
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:188]:
```suggestion
    out : {dpnp.ndarrays, usm_ndarray, None}, optional
```

File: dpnp/dpnp_iface_linearalgebra.py [hunk 7]
```diff
@@ -166,15 +167,59 @@ def dot(a, b, out=None):
     return dpnp.get_result_array(result, out, casting="no")
 
 
-def einsum(*args, **kwargs):
+def einsum(*operands, out=None, optimize=False, **kwargs):
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:170]: Should not we use the exact keyword rather then `kwargs` here?
```suggestion
def einsum(*operands, out=None, dtype=None, order="K", casting="safe", optimize=False):
```

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 8]
```diff
@@ -476,6 +1554,215 @@ def dpnp_cross(a, b, cp, exec_q):
...
+    """Evaluates the Einstein summation convention on the operands."""
+
+    input_subscripts, output_subscript, operands = _parse_einsum_input(operands)
+    assert isinstance(input_subscripts, list)
+    assert isinstance(operands, list)
+
+    dpnp.check_supported_arrays_type(*operands, scalar_type=True)
+    arrays = []
+    for a in operands:
+        if isinstance(a, (dpnp_array, dpt.usm_ndarray)):
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:1569]:
```suggestion
        if dpnp.is_supported_array_type(a):
```

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py [hunk 9]
```diff
@@ -476,6 +1554,215 @@ def dpnp_cross(a, b, cp, exec_q):
...
+                    )
+
+    _einsum_diagonals(input_subscripts, operands)
+
+    # no more raises
+    if len(operands) >= 2:
+        if any(arr.size == 0 for arr in operands):
+            return dpnp.zeros(
+                tuple(dimension_dict[label] for label in output_subscript),
+                dtype=result_dtype,
```
Comment (antonwolfy) [inline dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:1648]:
```suggestion
                dtype=result_dtype, usm_type=res_usm_type, sycl_queue=exec_q,
```

File: dpnp/dpnp_utils/dpnp_utils_linearalgebra.py (hunk 3 above)
Comment (vtavana) [reply dpnp/dpnp_utils/dpnp_utils_linearalgebra.py:870]: It seems this line is not needed, based on the logic of the code `len(full_result)` will always be at least 1, so removing it.

File: dpnp/dpnp_iface_linearalgebra.py [hunk 10]
```diff
@@ -183,36 +230,323 @@ def einsum(*args, **kwargs):
...
+           [13, 40, 67, 94]])
+    >>> np.einsum("ki,...k->i...", a, b)
+    array([[10, 28, 46, 64],
+           [13, 40, 67, 94]])
+    >>> np.einsum("k...,jk", a, b)
+    array([[10, 28, 46, 64],
+           [13, 40, 67, 94]])
+
+    Chained array operations. For more complicated contractions, speed ups
+    might be achieved by repeatedly computing a "greedy" path or pre-computing
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:382]: produce a spell warning:

> Spell check: pre: might be achieved by repeatedly computing a "greedy" path or pre-computing.

File: dpnp/dpnp_iface_linearalgebra.py [hunk 11]
```diff
@@ -183,36 +230,323 @@ def einsum(*args, **kwargs):
...
+    the "optimal" path and repeatedly applying it, using an `einsum_path`
+    insertion. Performance improvements can be particularly significant with
+    larger arrays:
+
+    >>> a = np.ones(64000).reshape(20, 40, 80)
+
+    Basic `einsum`: 119 ms ± 26 ms per loop (benchmarked on 12th
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:389]: spell warning:

> Spell check: benchmarked: : 119 ms ± 26 ms per loop (benchmarked on 12th

File: dpnp/dpnp_iface_linearalgebra.py [hunk 12]
```diff
@@ -183,36 +230,323 @@ def einsum(*args, **kwargs):
...
+        Choose the type of path. If a tuple is provided, the second argument is
+        assumed to be the maximum intermediate size created. If only a single
+        argument is provided the largest input or output array size is used
+        as a maximum intermediate size.
+
+        * if a list is given that starts with ``einsum_path``, uses this as the
+          contraction path
+        * if ``False`` or ``None`` no optimization is taken
+        * if ``True`` defaults to the "greedy" algorithm
+        * ``"optimal"`` is an algorithm that combinatorially explores all
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:451]: It's required to extend `doc/known_words.txt` with `combinatorially` word or rephrase.

File: dpnp/dpnp_iface_linearalgebra.py [hunk 13]
```diff
@@ -183,36 +230,323 @@ def einsum(*args, **kwargs):
...
+          possible ways of contracting the listed tensors and chooses the
+          least costly path. Scales exponentially with the number of terms
+          in the contraction.
+        * ``"greedy"`` is an algorithm that chooses the best pair contraction
+          at each step. Effectively, this algorithm searches the largest inner,
+          Hadamard, and then outer products at each step. Scales cubically with
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_linearalgebra.py:457]: It's required to extend doc/known_words.txt with `Hadamard` and `cubically`

**Review by antonwolfy — APPROVED** [review]
Thank you @vtavana

**Pull request merged**